builtInFunctions/setGuardian.go

```diff
@@ -0,0 +1,255 @@
+		mutExecution:             sync.RWMutex{},
+		keyPrefix:                []byte(core.ElrondProtectedKeyPrefix + GuardiansKeyIdentifier),
+	}
+	setGuardianFunc.baseEnabled = &baseEnabled{
+		function:        BuiltInFunctionSetGuardian,
+		activationEpoch: args.SetGuardianEnableEpoch,
+		flagActivated:   atomic.Flag{},
+	}
+
+	logAccountFreezer.Debug("set guardian enable epoch", args.SetGuardianEnableEpoch)
+	args.EpochNotifier.RegisterNotifyHandler(setGuardianFunc)
+
+	return setGuardianFunc, nil
+}
+
+// ProcessBuiltinFunction will process the set guardian built-in function call
+// Currently, the following cases are treated:
+// Case 1. User does NOT have any guardian => set guardian
+// Case 2. User has ONE guardian pending => do not set new guardian, wait until first one is set
+// Case 3. User has ONE guardian enabled => set guardian
+// Case 4. User has TWO guardians. FIRST is enabled, SECOND is pending => does not set, wait until second one is set
+// Case 5. User has TWO guardians. FIRST is enabled, SECOND is enabled => replace oldest one + set new one as pending
+func (sg *setGuardian) ProcessBuiltinFunction(
+	acntSnd, acntDst vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+	sg.mutExecution.RLock()
+	defer sg.mutExecution.RUnlock()
+
+	err := sg.checkArguments(acntSnd, acntDst, vmInput)
+	if err != nil {
+		return nil, err
+	}
+
+	guardians, err := sg.guardians(acntSnd)
+	if err != nil {
+		return nil, err
+	}
+	if sg.contains(guardians, vmInput.Arguments[0]) {
+		return nil, ErrGuardianAlreadyExists
+	}
+
+	switch len(guardians.Data) {
+	case 0:
+		// Case 1
+		return sg.tryAddGuardian(acntSnd, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 1:
+		// Case 2
+		if sg.pending(guardians.Data[0]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, hex.EncodeToString(guardians.Data[0].Address))
+		}
+		// Case 3
+		return sg.tryAddGuardian(acntSnd, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 2:
+		// Case 4
+		if sg.pending(guardians.Data[1]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, hex.EncodeToString(guardians.Data[1].Address))
+		}
+		// Case 5
+		guardians.Data = guardians.Data[1:] // remove oldest guardian
+		return sg.tryAddGuardian(acntSnd, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	default:
+		return &vmcommon.VMOutput{ReturnCode: vmcommon.ExecutionFailed}, nil
+	}
+}
+
+func (sg *setGuardian) checkArguments(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
+		return fmt.Errorf("%w for receiver", ErrNilUserAccount)
+	}
+	if vmInput == nil {
+		return ErrNilVmInput
+	}
+	if !(bytes.Equal(senderAccount.AddressBytes(), receiverAccount.AddressBytes()) &&
```

Comment:
> can you extract each conditional into a variable and use those instead?

Sure thing, did it for this condition. However, I'm not sure if it would bring much benefit doing it for all other conditions as well.

> Also shouldn't this condition be senderIsNotReceiver || senderIsNotCaller instead of senderIsNotReceiver && senderIsCaller

These conditions shall be equivalent:
`
!senderIsReceiver || !senderIsCaller <=> ! (senderIsReceiver && senderIsCaller)
`

I've decided to stick with the second one